### PR TITLE
Backtracking

### PR DESCRIPTION
**Pull request by AHarishma**
Add sudoku solver, unique paths, palindrome partitioning

### CODE REVIEW

File: src/com.company/backtracking/PalindromePartitioning.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package backtracking;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PalindromePartitioning {
+
+    void palindromeParitioning(String string, int index, List<String> currentList, List<List<String>> result) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/PalindromePartitioning.java]: Add access specifier for all the methods.

File: src/com.company/backtracking/PalindromePartitioning.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        if(index >= string.length()) {
+            result.add(new ArrayList<>(currentList));
+            return;
+        }
+
+        for(int i = index; i< string.length(); i++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/PalindromePartitioning.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/PalindromePartitioning.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            if(isPalindrome(string.substring(index, i + 1))) {
+                currentList.add(string.substring(index, i + 1));
+                palindromeParitioning(string, index + 1, currentList, result);
+                currentList.remove(currentList.size() - 1);
+            }
+        }
+    }
+
+    boolean isPalindrome(String string) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/PalindromePartitioning.java]: Add access specifier for all the methods.

File: src/com.company/backtracking/PalindromePartitioning.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        StringBuilder builder = new StringBuilder(string).reverse();
+        return string.equals(builder.toString());
+    }
+    public static void main(String[] args) {
+        List<List<String>> result = new ArrayList<>();
+        new PalindromePartitioning().palindromeParitioning("aabbc", 0, new ArrayList<>(), result);
+        for(int i= 0; i< result.size(); i++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/PalindromePartitioning.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/SudokuSolver.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            return true;
+        }
+        if(col == board[row].length) {
+            return canBeSolved(board, row + 1, 0);
+        }
+        if(board[row][col] != 0) {
+            return canBeSolved(board, row, col + 1);
+        }
+
+        for(int i = 1;  i<=9; i++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/SudokuSolver.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/SudokuSolver.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                    return true;
+                }
+                board[row][col] = 0;
+            }
+        }
+        return false;
+    }
+
+    private boolean isValidNumber(int[][] board, int row, int col, int number) {
+        for(int i = 0; i< board.length; i++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/SudokuSolver.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/SudokuSolver.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        for(int i = 0; i< board.length; i++) {
+            if(board[i][col] == number) {
+                return false;
+            }
+        }
+
+        int subGridRow = (row / 3) * 3;
+        int subGridCol = (col / 3) * 3;
+        for(int i = subGridRow;  i< subGridRow + 3; i++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/SudokuSolver.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/SudokuSolver.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            for(int j = subGridCol;  j< subGridCol + 3; j++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/SudokuSolver.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/UniquePaths.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
+package backtracking;
+
+import java.util.*;
+
+public class UniquePaths {
+
+    int[] rowIndex =  new int[] {-1, 0, 0, 1};
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide access specifier as private.

File: src/com.company/backtracking/UniquePaths.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    int[] colIndex = new int[]{0, -1, 1, 0};
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java:8]: Provide access specifier as private.

File: src/com.company/backtracking/UniquePaths.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    static int count = 0;
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide access specifier as private and remove static.

File: src/com.company/backtracking/UniquePaths.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    private int solve(int[][] maze) {
+        int[] start = new int[] {-1, -1};
+        for(int i = 0; i< maze.length; i++){
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/UniquePaths.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+            for(int j = 0; j< maze[0].length; j++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/UniquePaths.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            if(row>=0 &&  row < maze.length && col >= 0 && col < maze[row].length &&
+                    (maze[row][col] == 0 || maze[row][col] == 2) && !visited[row][col]) {
+                visited[row][col] = true;
+                dfs(maze, row, col, visited);
+                visited[row][col] = false;
+            }
+        }
+    }
+
+    boolean isAllVisited(int[][] maze, boolean[][] visited)  {
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide access specifier as private

File: src/com.company/backtracking/UniquePaths.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        for(int i = 0; i< maze.length; i++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide proper alignment. Use command + shift + enter.

File: src/com.company/backtracking/UniquePaths.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+            for(int j = 0; j< maze[0].length; j++) {
```
Comment (venkatyuva) [inline src/com.company/backtracking/UniquePaths.java]: Provide proper alignment. Use command + shift + enter.

**Review by venkatyuva — CHANGES_REQUESTED** [review]
Please correct the changes.

File: src/com.company/backtracking/PalindromePartitioning.java (hunk 1 above)
Comment (AHarishma) [reply src/com.company/backtracking/PalindromePartitioning.java]: added for all the methods

File: src/com.company/backtracking/PalindromePartitioning.java (hunk 2 above)
Comment (AHarishma) [reply src/com.company/backtracking/PalindromePartitioning.java]: formatted the code

File: src/com.company/backtracking/PalindromePartitioning.java (hunk 3 above)
Comment (AHarishma) [reply src/com.company/backtracking/PalindromePartitioning.java]: added for all the methods

File: src/com.company/backtracking/PalindromePartitioning.java (hunk 4 above)
Comment (AHarishma) [reply src/com.company/backtracking/PalindromePartitioning.java]: formatted the code

File: src/com.company/backtracking/SudokuSolver.java (hunk 5 above)
Comment (AHarishma) [reply src/com.company/backtracking/SudokuSolver.java]: formatted the code

File: src/com.company/backtracking/UniquePaths.java (hunk 12 above)
Comment (AHarishma) [reply src/com.company/backtracking/UniquePaths.java]: formatted the code

File: src/com.company/backtracking/UniquePaths.java (hunk 13 above)
Comment (AHarishma) [reply src/com.company/backtracking/UniquePaths.java]: formatted the code

File: src/com.company/backtracking/UniquePaths.java (hunk 15 above)
Comment (AHarishma) [reply src/com.company/backtracking/UniquePaths.java]: formatted the code

File: src/com.company/backtracking/UniquePaths.java (hunk 16 above)
Comment (AHarishma) [reply src/com.company/backtracking/UniquePaths.java]: formatted the code

File: src/com.company/backtracking/UniquePaths.java (hunk 14 above)
Comment (AHarishma) [reply src/com.company/backtracking/UniquePaths.java]: added for all methods

**Review by venkatyuva — APPROVED** [review]
Looks good to me.

**Pull request merged**